// fluid-music/built/techniques/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var basic_1 = require("./basic");
Object.defineProperty(exports, "PluginAutomation", { enumerable: true, get: function () { return basic_1.PluginAutomation; } });
Object.defineProperty(exports, "TrackAutomation", { enumerable: true, get: function () { return basic_1.TrackAutomation; } });
Object.defineProperty(exports, "ILayers", { enumerable: true, get: function () { return basic_1.ILayers; } });
Object.defineProperty(exports, "MidiChord", { enumerable: true, get: function () { return basic_1.MidiChord; } });
Object.defineProperty(exports, "MidiNote", { enumerable: true, get: function () { return basic_1.MidiNote; } });
Object.defineProperty(exports, "Random", { enumerable: true, get: function () { return basic_1.Random; } });
var audio_file_1 = require("./audio-file");
Object.defineProperty(exports, "AudioFile", { enumerable: true, get: function () { return audio_file_1.AudioFile; } });
Object.defineProperty(exports, "AFOnset", { enumerable: true, get: function () { return audio_file_1.AFOnset; } });
Object.defineProperty(exports, "AFReverse", { enumerable: true, get: function () { return audio_file_1.AFReverse; } });
Object.defineProperty(exports, "AFReverseLeadIn", { enumerable: true, get: function () { return audio_file_1.AFReverseLeadIn; } });
// Charles: It would be nice to export techniques like this:
// export * from './somewhere'
// Unfortunately, typedoc does not support this format, so we must remember to
// manually add export techniques
