// fluid-music/built/plugin-adapters/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var podolski_vst2_1 = require("./podolski-vst2");
Object.defineProperty(exports, "PodolskiVst2", { enumerable: true, get: function () { return podolski_vst2_1.PodolskiVst2; } });
exports.podolskiVst2Presets = require("./podolski-vst2-presets");
var rough_rider_3_vst2_1 = require("./rough-rider-3-vst2");
Object.defineProperty(exports, "RoughRider3Vst2", { enumerable: true, get: function () { return rough_rider_3_vst2_1.RoughRider3Vst2; } });
var t_compressor_vst2_1 = require("./t-compressor-vst2");
Object.defineProperty(exports, "TCompressorVst2", { enumerable: true, get: function () { return t_compressor_vst2_1.TCompressorVst2; } });
var t_equalizer_vst2_1 = require("./t-equalizer-vst2");
Object.defineProperty(exports, "TEqualizerVst2", { enumerable: true, get: function () { return t_equalizer_vst2_1.TEqualizerVst2; } });
var t_stereo_delay_vst2_1 = require("./t-stereo-delay-vst2");
Object.defineProperty(exports, "TStereoDelayVst2", { enumerable: true, get: function () { return t_stereo_delay_vst2_1.TStereoDelayVst2; } });
var tyrell_n6_vst2_1 = require("./tyrell-n6-vst2");
Object.defineProperty(exports, "TyrellN6Vst2", { enumerable: true, get: function () { return tyrell_n6_vst2_1.TyrellN6Vst2; } });
var dragonfly_room_vst2_1 = require("./dragonfly-room-vst2");
Object.defineProperty(exports, "DragonflyRoomVst2", { enumerable: true, get: function () { return dragonfly_room_vst2_1.DragonflyRoomVst2; } });
