// fluid-music/built/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UdpClient = void 0;
var FluidPlugin_1 = require("./FluidPlugin");
Object.defineProperty(exports, "FluidPlugin", { enumerable: true, get: function () { return FluidPlugin_1.FluidPlugin; } });
Object.defineProperty(exports, "PluginType", { enumerable: true, get: function () { return FluidPlugin_1.PluginType; } });
var sessionToFluidMessage_1 = require("./sessionToFluidMessage");
Object.defineProperty(exports, "sessionToContentFluidMessage", { enumerable: true, get: function () { return sessionToFluidMessage_1.sessionToContentFluidMessage; } });
Object.defineProperty(exports, "sessionToTemplateFluidMessage", { enumerable: true, get: function () { return sessionToFluidMessage_1.sessionToTemplateFluidMessage; } });
var sessionToReaperProject_1 = require("./sessionToReaperProject");
Object.defineProperty(exports, "sessionToReaperProject", { enumerable: true, get: function () { return sessionToReaperProject_1.sessionToReaperProject; } });
var FluidSession_1 = require("./FluidSession");
Object.defineProperty(exports, "FluidSession", { enumerable: true, get: function () { return FluidSession_1.FluidSession; } });
var FluidAudioFile_1 = require("./FluidAudioFile");
Object.defineProperty(exports, "FluidAudioFile", { enumerable: true, get: function () { return FluidAudioFile_1.FluidAudioFile; } });
Object.defineProperty(exports, "AudioFileMode", { enumerable: true, get: function () { return FluidAudioFile_1.AudioFileMode; } });
var FluidTrack_1 = require("./FluidTrack");
Object.defineProperty(exports, "FluidTrack", { enumerable: true, get: function () { return FluidTrack_1.FluidTrack; } });
exports.gen = require("./plugin-generator");
exports.converters = require("./converters");
exports.m = require("./m");
exports.random = require("./random");
exports.tLibrary = require("./t-library");
exports.tab = require("./tab");
// Unfortunately, JSDoc imports cannot currently access properties of namespaces
// that were exported via the `export * as namespace from './somewhere'` format.
// In order to access techniques, JSDoc imports have to import from the build
// directory, which is likely to change in the future.
//
// While while the example below is unstable, it is also (currently) the only
// option for non-typescript users:
// /** @typedef {import('fluid-music/built/techniques').AudioFile} AudioFile */
//
// Update: it appears that `export * from './somewhere'` is not well supported
// by typedoc. For that reason, I'm manually exporting names in techniques
// as opposed to using *. This probably means that it is possible to import
// techniques with the nicer syntax (see UseContext example below).
exports.techniques = require("./techniques");
// New Style Plugins
exports.plugins = require("./plugin-adapters/index");
// OSC Message Helpers
exports.cybr = require("./cybr/index");
exports.UdpClient = require('./cybr/UdpClient');
var IpcClient_1 = require("./cybr/IpcClient");
Object.defineProperty(exports, "IpcClient", { enumerable: true, get: function () { return IpcClient_1.IpcClient; } });
var IpcClient_2 = require("./cybr/IpcClient");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return IpcClient_2.IpcClient; } });
