// fluid-music/built/m.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.m = void 0;
/**
 * Map from note names with octaves to midi pitch numbers. Ex:
 * c: 0,
 * c1: 12,
 * c4: 60,
 * a4: 69,
 * ...etc...
 */
exports.m = {
    c: 0,
    cs: 1,
    df: 1,
    d: 2,
    ds: 3,
    ef: 3,
    e: 4,
    f: 5,
    fs: 6,
    gf: 6,
    g: 7,
    gs: 8,
    af: 8,
    a: 9,
    as: 10,
    bf: 10,
    b: 11,
    c0: 12,
    cs0: 13,
    df0: 13,
    d0: 14,
    ds0: 15,
    ef0: 15,
    e0: 16,
    f0: 17,
    fs0: 18,
    gf0: 18,
    g0: 19,
    gs0: 20,
    af0: 20,
    a0: 21,
    as0: 22,
    bf0: 22,
    b0: 23,
    c1: 24,
    cs1: 25,
    df1: 25,
    d1: 26,
    ds1: 27,
    ef1: 27,
    e1: 28,
    f1: 29,
    fs1: 30,
    gf1: 30,
    g1: 31,
    gs1: 32,
    af1: 32,
    a1: 33,
    as1: 34,
    bf1: 34,
    b1: 35,
    c2: 36,
    cs2: 37,
    df2: 37,
    d2: 38,
    ds2: 39,
    ef2: 39,
    e2: 40,
    f2: 41,
    fs2: 42,
    gf2: 42,
    g2: 43,
    gs2: 44,
    af2: 44,
    a2: 45,
    as2: 46,
    bf2: 46,
    b2: 47,
    c3: 48,
    cs3: 49,
    df3: 49,
    d3: 50,
    ds3: 51,
    ef3: 51,
    e3: 52,
    f3: 53,
    fs3: 54,
    gf3: 54,
    g3: 55,
    gs3: 56,
    af3: 56,
    a3: 57,
    as3: 58,
    bf3: 58,
    b3: 59,
    c4: 60,
    cs4: 61,
    df4: 61,
    d4: 62,
    ds4: 63,
    ef4: 63,
    e4: 64,
    f4: 65,
    fs4: 66,
    gf4: 66,
    g4: 67,
    gs4: 68,
    af4: 68,
    a4: 69,
    as4: 70,
    bf4: 70,
    b4: 71,
    c5: 72,
    cs5: 73,
    df5: 73,
    d5: 74,
    ds5: 75,
    ef5: 75,
    e5: 76,
    f5: 77,
    fs5: 78,
    gf5: 78,
    g5: 79,
    gs5: 80,
    af5: 80,
    a5: 81,
    as5: 82,
    bf5: 82,
    b5: 83,
    c6: 84,
    cs6: 85,
    df6: 85,
    d6: 86,
    ds6: 87,
    ef6: 87,
    e6: 88,
    f6: 89,
    fs6: 90,
    gf6: 90,
    g6: 91,
    gs6: 92,
    af6: 92,
    a6: 93,
    as6: 94,
    bf6: 94,
    b6: 95,
    c7: 96,
    cs7: 97,
    df7: 97,
    d7: 98,
    ds7: 99,
    ef7: 99,
    e7: 100,
    f7: 101,
    fs7: 102,
    gf7: 102,
    g7: 103,
    gs7: 104,
    af7: 104,
    a7: 105,
    as7: 106,
    bf7: 106,
    b7: 107,
    c8: 108,
    cs8: 109,
    df8: 109,
    d8: 110,
    ds8: 111,
    ef8: 111,
    e8: 112,
    f8: 113,
    fs8: 114,
    gf8: 114,
    g8: 115,
    gs8: 116,
    af8: 116,
    a8: 117,
    as8: 118,
    bf8: 118,
    b8: 119,
    c9: 120,
    cs9: 121,
    df9: 121,
    d9: 122,
    ds9: 123,
    ef9: 123,
    e9: 124,
    f9: 125,
    fs9: 126,
    gf9: 126,
    g9: 127,
};
exports.default = exports.m;
